### PR TITLE
Feature/dengage token

### PR DESCRIPTION
**Pull request by erdem-burgan**
Developed features for Dengage to take token firabase and set in Dengage-SDK.
Catch  click event of Dengage notification and grap its target url and then use for deeplink navigation.

## Release Notes

- **New Features**
  - Introduced `DengageMessage` and `DengageMedia` classes for structured messaging data handling.
  - Enhanced notification handling with token management in Firebase and Huawei messaging.
  - Added state management and event handling capabilities to the `NeoCoreMessaging` class.

- **Bug Fixes**
  - Improved error handling and logging for notification events.

- **Documentation**
  - Updated constructor signatures for messaging classes to include required parameters.

### CODE REVIEW

File: lib/core/widgets/models/dengage_message.dart [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    required this.targetUrl,
+    required this.title,
+    required this.transactionId,
+  });
+
+  factory DengageMessage.fromJson(Map<String, dynamic> json) => _$DengageMessageFromJson(json);
+}
+
+@JsonSerializable(createToJson: false)
+class Media {
```
Comment (reyhanlioglu) [inline lib/core/widgets/models/dengage_message.dart:81]: buna DengageMessageMedia desek daha iyi olabilir abi fazla genel bi isimlendirme gibi, karışmasın ilerde gelebilecek başka modellerle

File: lib/core/widgets/neo_core_firebase_messaging/neo_core_firebase_messaging.dart [hunk 2]
```diff
@@ -35,6 +37,7 @@ class NeoCoreFirebaseMessaging extends StatefulWidget {
   final Widget child;
   final NeoNetworkManager networkManager;
   final NeoCoreSecureStorage neoCoreSecureStorage;
+  final Function(String) token;
```
Comment (reyhanlioglu) [inline lib/core/widgets/neo_core_firebase_messaging/neo_core_firebase_messaging.dart:40]: onTokenChanged diyebiliriz sanki function ismine

File: lib/core/widgets/neo_core_messaging/neo_core_messaging.dart [hunk 3]
```diff
@@ -22,35 +28,69 @@ class NeoCoreMessaging extends StatelessWidget {
...
+    try {
+      final Map<String, dynamic> eventData = json.decode(event);
+      final dengageMessage = DengageMessage.fromJson(eventData);
+      if (_Constants.messageSource.toLowerCase() == dengageMessage.messageSource.toLowerCase() &&
+          dengageMessage.media.isNotEmpty &&
+          dengageMessage.media[0].target.isNotEmpty) {
+        widget.onDeeplinkNavigation?.call(dengageMessage.media[0].target);
+      }
+    } catch (e) {
+      debugPrint("Dengage Message Error is: $e");
```
Comment (reyhanlioglu) [inline lib/core/widgets/neo_core_messaging/neo_core_messaging.dart:54]: hataları NeoLogger ile loglayabiliriz

File: lib/core/widgets/neo_core_messaging/neo_core_messaging.dart [hunk 4]
```diff
@@ -1,19 +1,25 @@
...
+abstract class _Constants {
+  static const messageSource = "DENGAGE";
+}
+
+class NeoCoreMessaging extends StatefulWidget {
   final Widget child;
   final NeoSharedPrefs neoSharedPrefs;
   final NeoNetworkManager networkManager;
   final NeoCoreSecureStorage neoCoreSecureStorage;
+  final Function(String) firebaseToken;
```
Comment (reyhanlioglu) [inline lib/core/widgets/neo_core_messaging/neo_core_messaging.dart:22]: buraya da onTokenChanged diyebiliriz

File: lib/core/widgets/neo_core_messaging/neo_core_messaging.dart [hunk 5]
```diff
@@ -22,35 +28,69 @@ class NeoCoreMessaging extends StatelessWidget {
...
         );
       }
     }
   }
+
+  @override
+  void initState() {
+    eventChannel.receiveBroadcastStream().listen(_onEvent, onError: _onError);
+    super.initState();
+  }
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_core_messaging/neo_core_messaging.dart:95]: bot +1, hatta en üstte yazsak daha okunaklı olur initState methodunu üstlere dispose'u en alta yazıyouz genellikle

[the remark this answers is not in the document]
Comment (erdem-burgan) [reply lib/core/widgets/neo_core_messaging/neo_core_messaging.dart:95]: I'll take account your comments I am going to fix it according to them

File: lib/core/widgets/models/dengage_message.dart [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  @JsonKey(name: "dengageSendId", defaultValue: 0)
+  final int dengageSendId;
+  @JsonKey(name: "sound", defaultValue: "")
+  final String sound;
+  @JsonKey(name: "source", defaultValue: "")
+  final String source;
+  @JsonKey(name: "subTitle", defaultValue: "")
+  final String subTitle;
+  @JsonKey(name: "targetUrl", defaultValue: "")
+  final String targetUrl;
```
[the remark this answers is not in the document]
Comment (erdem-burgan) [reply lib/core/widgets/models/dengage_message.dart:46]: we are taking target data intentionally from that part. we tested it. thank you.

**Review by reyhanlioglu — APPROVED** [review]

**Review by alperen-commencis — APPROVED** [review]

**Review by alperen-commencis — APPROVED** [review]

**Review by alperen-commencis — APPROVED** [review]

**Review by mcanyigincommencis — APPROVED** [review]

**Pull request merged**